File: app/lib/shared/models/user_model.dart
class UserModel {
  final String id;
  final String name;
  final String? username;
  final String email;
  final String? phone;
  final String? avatarUrl;
  final int streakCurrent;
  final int streakBest;
  final int pfTotal;
  final int pfToNextLevel;
  final int pfWeekly;
  final int level;
  final int energy;
  final int manas;
  final String? denomination;
  final int dailyGoalMinutes;
  final String? fcmToken;
  final String? leagueTier;
  final int? leagueRank;

  const UserModel({
    required this.id,
    required this.name,
    this.username,
    required this.email,
    this.phone,
    this.avatarUrl,
    required this.streakCurrent,
    required this.streakBest,
    required this.pfTotal,
    this.pfToNextLevel = 1000,
    required this.pfWeekly,
    required this.level,
    this.energy = 20,
    required this.manas,
    this.denomination,
    required this.dailyGoalMinutes,
    this.fcmToken,
    this.leagueTier,
    this.leagueRank,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    final pfTotal = (json['pf_total'] ?? json['xp_total'] ?? 0) as int;
    // Compute level from pfTotal (1000 PF per level) — override stale DB value
    final computedLevel = pfTotal ~/ 1000 + 1;
    final storedLevel = (json['level'] ?? 1) as int;

    return UserModel(
      id: json['_id'] ?? json['id'] ?? '',
      name: json['name'] ?? '',
      username: json['username'],
      email: json['email'] ?? '',
      phone: json['phone'],
      avatarUrl: json['avatar_url'],
      streakCurrent: json['streak_current'] ?? 0,
      streakBest: json['streak_best'] ?? json['streak_longest'] ?? 0,
      pfTotal: pfTotal,
      pfToNextLevel: 1000,
      pfWeekly: json['pf_weekly'] ?? json['xp_weekly'] ?? 0,
      level: computedLevel > storedLevel ? computedLevel : storedLevel,
      energy: json['energy'] ?? json['hearts'] ?? 20,
      manas: json['manas'] ?? 0,
      denomination: json['denomination'],
      dailyGoalMinutes: json['daily_goal_minutes'] ?? 10,
      fcmToken: json['fcm_token'],
      leagueTier: json['league_tier'],
      leagueRank: json['league_rank'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'username': username,
      'email': email,
      'phone': phone,
      'avatar_url': avatarUrl,
      'streak_current': streakCurrent,
      'streak_best': streakBest,
      'pf_total': pfTotal,
      'pf_to_next_level': pfToNextLevel,
      'pf_weekly': pfWeekly,
      'level': level,
      'energy': energy,
      'manas': manas,
      'denomination': denomination,
      'daily_goal_minutes': dailyGoalMinutes,
      'fcm_token': fcmToken,
      'league_tier': leagueTier,
      'league_rank': leagueRank,
    };
  }

  UserModel copyWith({
    String? id,
    String? name,
    String? username,
    String? email,
    String? phone,
    String? avatarUrl,
    int? streakCurrent,
    int? streakBest,
    int? pfTotal,
    int? pfToNextLevel,
    int? pfWeekly,
    int? level,
    int? energy,
    int? manas,
    String? denomination,
    int? dailyGoalMinutes,
    String? fcmToken,
    String? leagueTier,
    int? leagueRank,
  }) {
    return UserModel(
      id: id ?? this.id,
      name: name ?? this.name,
      username: username ?? this.username,
      email: email ?? this.email,
      phone: phone ?? this.phone,
      avatarUrl: avatarUrl ?? this.avatarUrl,
      streakCurrent: streakCurrent ?? this.streakCurrent,
      streakBest: streakBest ?? this.streakBest,
      pfTotal: pfTotal ?? this.pfTotal,
      pfToNextLevel: pfToNextLevel ?? this.pfToNextLevel,
      pfWeekly: pfWeekly ?? this.pfWeekly,
      level: level ?? this.level,
      energy: energy ?? this.energy,
      manas: manas ?? this.manas,
      denomination: denomination ?? this.denomination,
      dailyGoalMinutes: dailyGoalMinutes ?? this.dailyGoalMinutes,
      fcmToken: fcmToken ?? this.fcmToken,
      leagueTier: leagueTier ?? this.leagueTier,
      leagueRank: leagueRank ?? this.leagueRank,
    );
  }
}

File: api/src/modules/users/user.model.ts
import mongoose, { Schema, Document } from 'mongoose';
import { UserPlan, LeagueTier } from '../../shared/types';

export interface IUser extends Document {
  name: string;
  username: string;
  email: string;
  phone?: string;
  passwordHash: string;
  avatar_url?: string;
  
  // Plano
  plan: UserPlan;
  plan_expires_at?: Date;
  
  // Gamificação
  pf_total: number;
  pf_weekly: number;
  level: number;
  pf_to_next_level: number;
  energy: number;
  energy_last_consumed: Date;
  manas: number;
  
  // Streak
  streak_current: number;
  streak_longest: number;
  streak_last_activity: Date;
  streak_freeze_count: number;
  
  // Liga
  league_id?: mongoose.Types.ObjectId;
  league_tier: LeagueTier;
  
  // Perfil
  denomination?: string;
  daily_goal_minutes: number;
  timezone: string;
  
  // Push
  fcm_token?: string;
  notifications_enabled: boolean;
  notification_hour: number;
  
  // Institutional
  company_id?: mongoose.Types.ObjectId;
  
  // Status
  is_active: boolean;
  last_seen_at: Date;
  role: 'user' | 'admin' | 'company_admin';
  createdAt: Date;
  updatedAt: Date;
}

const UserSchema = new Schema<IUser>(
  {
    name: { type: String, required: true, trim: true },
    username: { type: String, required: true, unique: true, trim: true },
    email: { type: String, required: true, unique: true, lowercase: true, trim: true },
    phone: { type: String, trim: true },
    passwordHash: { type: String, required: true, select: false },
    avatar_url: { type: String },
    
    plan: { type: String, enum: ['free', 'plus_monthly', 'plus_annual'], default: 'free' },
    plan_expires_at: { type: Date },
    
    pf_total: { type: Number, default: 0 },
    pf_weekly: { type: Number, default: 0 },
    level: { type: Number, default: 1 },
    pf_to_next_level: { type: Number, default: 1000 },
    energy: { type: Number, default: 20, max: 20 },
    energy_last_consumed: { type: Date },
    manas: { type: Number, default: 200 },
    
    streak_current: { type: Number, default: 0 },
    streak_longest: { type: Number, default: 0 },
    streak_last_activity: { type: Date, default: Date.now },
    streak_freeze_count: { type: Number, default: 0 },
    
    league_id: { type: Schema.Types.ObjectId, ref: 'League' },
    league_tier: { type: String, enum: ['ruben', 'simeao', 'levi', 'juda', 'da', 'naftali', 'gad', 'aser', 'issacar', 'zebulom', 'efraim', 'manasses'], default: 'ruben' },
    
    denomination: { type: String },
    daily_goal_minutes: { type: Number, default: 10 },
    timezone: { type: String, default: 'America/Sao_Paulo' },
    
    fcm_token: { type: String },
    notifications_enabled: { type: Boolean, default: true },
    notification_hour: { type: Number, default: 20 },
    
    company_id: { type: Schema.Types.ObjectId, ref: 'Company' },
    role: { type: String, enum: ['user', 'admin', 'company_admin'], default: 'user' },
    
    is_active: { type: Boolean, default: true },
    last_seen_at: { type: Date, default: Date.now },
  },
  { timestamps: true }
);

UserSchema.index({ email: 1 });
UserSchema.index({ username: 1 });
UserSchema.index({ company_id: 1 });
UserSchema.index({ league_id: 1 });
UserSchema.index({ streak_last_activity: 1 });

export const UserModel = mongoose.model<IUser>('User', UserSchema);
